File: frontend/src/types/decision.ts
export type DecisionStateValue = 'NO_TARGET' | 'TRACKING' | 'WAIT' | 'LOCKED' | 'FIRE_READY' | 'NO_FIRE' | 'FAULT'
export type GateStatus = 'pass' | 'fail' | 'warning' | 'not_applicable'
export type GateSeverity = 'info' | 'warning' | 'critical'

export interface SafetyGate {
  name: string
  status: GateStatus
  severity: GateSeverity
  reason: string
  updated_at: number
}

export interface DecisionState {
  decision_state: DecisionStateValue
  fire_policy: string
  active_target_id: number | null
  selected_body_detection_id: number | null
  selected_balloon_detection_id: number | null
  target_class: string | null
  target_team: string
  range_m: number | null
  stable_frames: number
  required_stable_frames: number
  gates: SafetyGate[]
  blocking_reasons: string[]
  decision_reason: string
  updated_at: number
  aim_point: Record<string, unknown> | null
  person_safety: PersonSafetyGateStatus | null
}

export interface PersonSafetyGateStatus {
  enabled: boolean
  person_detected: boolean
  fire_gate_blocked_reason: string | null
  recommended_state: 'CLEAR' | 'SAFE_HOLD' | 'FIRE_BLOCKED'
  confidence_threshold: number
  hold_ms: number
  clear_after_ms: number
  last_detection_confidence: number | null
  last_detection_class: string | null
  last_detection_id: number | null
  last_detection_timestamp_ms: number | null
  active_until_ms: number | null
  source: string
  no_physical_command_generated: boolean
}

export interface FireEvaluationResult {
  accepted: boolean
  dry_run: boolean
  decision_state: DecisionStateValue
  blocking_reasons: string[]
  gates: SafetyGate[]
  reason: string
}

export interface ArmDisarmResult {
  accepted: boolean
  armed: boolean
  reason: string
  blocking_reasons: string[]
  decision: DecisionState | null
}

File: backend/app/schemas/decision.py
from enum import StrEnum
from typing import Any

from pydantic import BaseModel, Field

from app.schemas.person_safety import PersonSafetyGateStatus


class DecisionStateValue(StrEnum):
    NO_TARGET = "NO_TARGET"
    TRACKING = "TRACKING"
    WAIT = "WAIT"
    LOCKED = "LOCKED"
    FIRE_READY = "FIRE_READY"
    NO_FIRE = "NO_FIRE"
    FAULT = "FAULT"


class GateStatus(StrEnum):
    PASS = "pass"
    FAIL = "fail"
    WARNING = "warning"
    NOT_APPLICABLE = "not_applicable"


class GateSeverity(StrEnum):
    INFO = "info"
    WARNING = "warning"
    CRITICAL = "critical"


class SafetyGate(BaseModel):
    name: str
    status: GateStatus
    severity: GateSeverity
    reason: str
    updated_at: float


class DecisionState(BaseModel):
    decision_state: DecisionStateValue
    fire_policy: str = "NO_FIRE_DEFAULT"
    active_target_id: int | None = None
    selected_body_detection_id: int | None = None
    selected_balloon_detection_id: int | None = None
    target_class: str | None = None
    target_team: str = "unknown"
    range_m: float | None = None
    stable_frames: int = 0
    required_stable_frames: int = 5
    gates: list[SafetyGate] = Field(default_factory=list)
    blocking_reasons: list[str] = Field(default_factory=list)
    decision_reason: str
    updated_at: float
    aim_point: dict[str, Any] | None = None
    person_safety: PersonSafetyGateStatus | None = None


class FireEvaluationResult(BaseModel):
    accepted: bool
    dry_run: bool
    decision_state: DecisionStateValue
    blocking_reasons: list[str]
    gates: list[SafetyGate]
    reason: str


class ArmDisarmResult(BaseModel):
    accepted: bool
    armed: bool
    reason: str
    blocking_reasons: list[str] = Field(default_factory=list)
    decision: DecisionState | None = None
